pages/_app.tsx

```diff
@@ -24,9 +26,48 @@ export default function App({ Component, pageProps }: AppProps) {
     >
       <UserInfoProvider>
         <Layout>
-          <Component {...pageProps} />
+          {pageProps.protected ? (
```

Comment:
Protected pages require user to be signed in and have the userinfo context defined, and premium pages require that and to have a premium membership. Conditionally sends to signin() if first case is not met, otherwise shows upgrade component if not premium member.